### PR TITLE
feat(urd): use `validAt` instead of `submittedAt`

### PR DESCRIPTION
**Pull request by julien-devatom**
Cantina issue: [cantinasec/review-morpho-blue-1#81](https://github.com/cantinasec/review-morpho-blue-1/issues/81)

### CODE REVIEW

File: src/UniversalRewardsDistributor.sol [hunk 1]
```diff
@@ -145,15 +145,8 @@ contract UniversalRewardsDistributor is IUniversalRewardsDistributorStaticTyping
     /// @notice Sets the timelock of a given distribution.
     /// @param newTimelock The new timelock.
     /// @dev This function can only be called by the owner of the distribution.
-    /// @dev If the timelock is reduced, it can only be updated after the timelock has expired.
+    /// @dev The timelock modification are not applicable to the pending values.
```
Comment (Rubilmax) [inline src/UniversalRewardsDistributor.sol:148]:
```suggestion
    /// @dev Changes to the timelock do not impact the timelock duration of the pending root.
```

**Review by Rubilmax — APPROVED** [review]
These are indeed the minimal changes to switch to a `validAt` mechanism, but the cleanest code will be when it'll get merged with #83 (resolving conflicts)

**Comment by julien-devatom** [comment]
> These are indeed the minimal changes to switch to a `validAt` mechanism, but the cleanest code will be when it'll get merged with #83 (resolving conflicts)

Merge day is going to be my worst day of the year

**Comment by Rubilmax** [comment]
> > These are indeed the minimal changes to switch to a `validAt` mechanism, but the cleanest code will be when it'll get merged with #83 (resolving conflicts)
> 
> Merge day is going to be my worst day of the year

You won't be doing it alone! Ask for our reviews for each PR btw

**Comment by julien-devatom** [comment]
> > > These are indeed the minimal changes to switch to a `validAt` mechanism, but the cleanest code will be when it'll get merged with #83 (resolving conflicts)
> > 
> > 
> > Merge day is going to be my worst day of the year
> 
> You won't be doing it alone! Ask for our reviews for each PR btw (I spotted at least 6 badly resolved merge conflicts on bundlers)

Yes ofc. I think this flow can be improved a lot.

**Comment by MerlinEgalite** [comment]
> > These are indeed the minimal changes to switch to a `validAt` mechanism, but the cleanest code will be when it'll get merged with #83 (resolving conflicts)
> 
> Merge day is going to be my worst day of the year

Naaah it'll be fine

File: test/UniversalRewardsDistributorTest.sol [hunk 2]
```diff
@@ -289,17 +289,34 @@ contract UniversalRewardsDistributorTest is Test {
         assertEq(distributionWithoutTimeLock.timelock(), newTimelock);
     }
 
-    function testSetTimelockShouldIncreaseTheQueueTimestamp(
-        uint256 timeElapsed,
-        uint256 newTimelock,
-        uint256 beforeEndOfTimelock,
-        uint256 afterEndOfTimelock
-    ) public {
+    function testSetTimelockShouldNotImpactPendingValuesIfTimelockIncreased(uint256 timeElapsed, uint256 newTimelock)
```
Comment (MerlinEgalite) [inline test/UniversalRewardsDistributorTest.sol:292]: to add in your PR @Rubilmax maybe

**Review by MerlinEgalite — APPROVED** [review]

File: test/UniversalRewardsDistributorTest.sol [hunk 3]
```diff
@@ -289,17 +289,34 @@ contract UniversalRewardsDistributorTest is Test {
...
+        public
+    {
+        newTimelock = bound(newTimelock, DEFAULT_TIMELOCK + 1, type(uint256).max - block.timestamp);
+        vm.assume(timeElapsed > DEFAULT_TIMELOCK && timeElapsed <= newTimelock);
```
Comment (Rubilmax) [inline test/UniversalRewardsDistributorTest.sol:296]: this could be bound to avoid discarding test cases too often :p

**Comment by MerlinEgalite** [comment]
Fixes #https://github.com/cantinasec/review-morpho-blue-1/issues/111 as well

File: src/UniversalRewardsDistributor.sol [hunk 4]
```diff
@@ -83,7 +83,7 @@ contract UniversalRewardsDistributor is IUniversalRewardsDistributorStaticTyping
         if (timelock == 0) {
             _setRoot(newRoot, newIpfsHash);
         } else {
-            pendingRoot = PendingRoot(block.timestamp, newRoot, newIpfsHash);
+            pendingRoot = PendingRoot(block.timestamp + timelock, newRoot, newIpfsHash);
             emit EventsLib.RootProposed(newRoot, newIpfsHash);
```
Comment (StErMi) [inline src/UniversalRewardsDistributor.sol:87]: would it make sense to also log when the pending proposal timelock will expire?

**Review by StErMi — APPROVED** [review]

File: src/UniversalRewardsDistributor.sol (hunk 4 above)
Comment (MerlinEgalite) [reply src/UniversalRewardsDistributor.sol:87]: There's no expiration on the URD, maybe you mixed it with MM?

Comment (StErMi) [reply src/UniversalRewardsDistributor.sol:87]: sorry, I meant when it would be "acceptable".

Comment (MerlinEgalite) [reply src/UniversalRewardsDistributor.sol:87]: In this case it's already the case since we log `block.timestamp + timelock` no?

Comment (StErMi) [reply src/UniversalRewardsDistributor.sol:87]: Where is such a log? The current event does not log such information.

Comment (MerlinEgalite) [reply src/UniversalRewardsDistributor.sol:87]: nevermind I mistake the code lines. @julien-devatom do you think it's necessary?

**Pull request merged**